blockcerts/blocksigning.go

```diff
@@ -0,0 +1,83 @@
+package blockcerts
+
+import (
+    "context"
+    "github.com/spacemeshos/go-spacemesh/blockcerts/types"
+    "github.com/spacemeshos/go-spacemesh/codec"
+    "github.com/spacemeshos/go-spacemesh/hare"
+    "github.com/spacemeshos/go-spacemesh/log"
+    "github.com/spacemeshos/go-spacemesh/p2p/pubsub"
+    "github.com/spacemeshos/go-spacemesh/signing"
+)
+
+func blockSigningLoop(ctx context.Context,
+    hareTerminationsCh <-chan hare.TerminationBlockOutput,               // input
+    blockSigner signing.Signer, committeeSize int, rolacle hare.Rolacle, // dependencies
+    gossipPublisher pubsub.Publisher, signatureCache *sigCache,          // output
+    logger log.Logger,
+) {
+    logger = logger.WithContext(ctx)
+    for {
+        select {
+        case <-ctx.Done():
+            return
+
+        case hareTermination, moreToCome := <-hareTerminationsCh:
+            if !moreToCome {
+                logger.Debug("block certification service: " +
+                    "hare termination channel closed.")
+                return
+            }
+            // 1. check readiness
+            ready, err := rolacle.IsIdentityActiveOnConsensusView(ctx,
+                hareTermination.TerminatingNodeID,
+                hareTermination.LayerID,
+            )
+            if err != nil {
+                logger.Error("error checking if active on consensus view")
+            }
+            if !ready {
+                logger.Debug("not active on consensus view: " +
+                    "not eligible to get committee seats.")
+                break
+            }
+            // 2. calculate eligibility
+            proof, err := rolacle.Proof(ctx,
+                hareTermination.LayerID, blockCertifierRole)
+            if err != nil {
+                logger.Error("could not retrieve eligibility proof "+
+                    "from oracle: %v", err)
+                break
+            }
+            committeeSeatCount, err := rolacle.CalcEligibility(ctx,
+                hareTermination.LayerID, blockCertifierRole, committeeSize,
+                hareTermination.TerminatingNodeID, proof)
+
+            blockIDBytes := hareTermination.BlockID.Bytes()
+            blockIDSignature := blockSigner.Sign(blockIDBytes)
```

Comment:
why only block id is signed? i would expect that whole message needs to be signed

Reply:
I was under the impression that the public key was the same for the VRF signer and the regular ed signer. 
https://github.com/spacemeshos/go-spacemesh/blob/eedff1e659de089ab4497c3614291b213f256479/signing/signer.go#L137

If this is the case, then it seems redundant to sign the whole message since every other field goes into the VRF verifier anyways.

Come to think of it, maybe we want early warning of tampering in transit so that we don't need to run the oracle validation if the msg signature is invalid: we can reject sooner. 
I guess it's also more idiomatic to validate a message than only the BlockID.
I'm convinced.

Reply:
i don't get your reasoning on rejecting sooner. 

the point of signing is either you can trust the whole msg is from the signer or not. for example, node X can use node Y's signature on blockID and provides bad proof or wrong layer id. how are you going to figure that out if you only sign blockID?

Reply:
Because node X has a different public key from node Y. 
You need to check that the regular ED signature matches the VRF signature public key whether you sign the whole message with EdSigner or just the BlockID.

The "rejecting or ignoring sooner" is just saving us the time spent for VRF proof validation in the case that some node signed the message with their public key but used some other node's VRF proof.

Reply:
hmmm... using your code now, i can get away with this without causing an error.
```
blockSigMsg.LayerID = bogusLayerID
blockSigMsg.SignerNodeID = Y.NodeID
blockSigMsg.SignerRoleProof = Y.proof
blockSigMsg.BlockID = hareTermination.BlockID
blockSigMsg.BlockIDSignature = X.blockIDSignature
```
in the gossip handler, you need to 
- extract the public key from the signed msg after you decode it
- call roracle.Validate() with the extract public key, not the msg.SignerNodeID
that way the above scenario will be rejected. 

> The "rejecting or ignoring sooner" is just saving us the time spent for VRF proof validation in the case that some node signed the message with their public key but used some other node's VRF proof.

no. when you extract the public key from the signed message, you have no idea yet whether the public key is valid or not until you do the VRF validate. there is no saving to be had.

Reply:
Yeah, this code doesn't check that yet. I was trying to find a second public key that didn't exist to make sure I wasn't missing something.

Oh, yeah I guess if we extract the public key from the signature, then we end up with nothing really useful because obviously the public key will match the signature. The only "early warning" we get is if we fail to extract a public key.

Reply:
Okay, so then I finally maintain that a signature of only the BlockID is required.

Reply:
say you are only signing BlockID, consider this:

BadBob receives a msg generated by from GoodAlice, changed `blockSigMsg.proof` to garbage, and relayed this gossip to InnocentJoe. when InnocentJoe sees that proof doesn't validate with GoodAlice's public key extracted from the blockID signature, it it marks GoodAlice malicious.